back/engines/commercial/analysis/app/services/analysis/inputs_finder.rb

```diff
@@ -0,0 +1,136 @@
+# frozen_string_literal: true
+
+module Analysis
+  class InputsFinder
+    MAX_PER_PAGE = 100
+
+    attr_reader :analysis, :params
+
+    def initialize(analysis, params = {})
+      @analysis = analysis
+      @params = params
+    end
+
+    def execute
+      inputs = analysis.inputs
+
+      inputs = filter_published_at(inputs)
+      inputs = filter_reactions(inputs)
+      inputs = filter_comments(inputs)
+      inputs = filter_votes(inputs)
+      inputs = filter_custom_field_in(inputs)
+      inputs = filter_custom_field_range(inputs)
+      search(inputs)
+    end
+
+    private
+
+    def filter_published_at(inputs)
+      scope = inputs
+
+      scope = scope.where('published_at >= ?', params[:published_at_from]) if params[:published_at_from]
+      scope = scope.where('published_at <= ?', params[:published_at_to]) if params[:published_at_to]
+
+      scope
+    end
+
+    def filter_reactions(inputs)
+      scope = inputs
+
+      scope = scope.where('dislikes_count + likes_count >= ?', params[:reactions_from]) if params[:reactions_from]
+      scope = scope.where('dislikes_count + likes_count <= ?', params[:reactions_to]) if params[:reactions_to]
+
+      scope
+    end
+
+    def filter_comments(inputs)
+      scope = inputs
+
+      scope = scope.where('comments_count >= ?', params[:comments_from]) if params[:comments_from]
+      scope = scope.where('comments_count <= ?', params[:comments_to]) if params[:comments_to]
+
+      scope
+    end
+
+    def filter_votes(inputs)
+      scope = inputs
+
+      scope = scope.where('votes_count >= ?', params[:votes_from]) if params[:votes_from]
+      scope = scope.where('votes_count <= ?', params[:votes_to]) if params[:votes_to]
+
+      scope
+    end
+
+    def search(inputs)
+      return inputs unless (search = params[:search])
+
+      inputs.search_by_all(search)
+    end
+
+    def filter_custom_field_in(inputs)
+      scope = inputs
+
+      decode_author_in_custom_keys.each do |(custom_field_id, value)|
+        raise ArgumentError, "value specified for author_custom_#{custom_field_id} must be an array" unless value.is_a? Array
+
+        cf = CustomField.find(custom_field_id)
+        case cf.input_type
+        when 'select', 'date'
+          scope = if value.empty?
```

Comment:
If we think in terms of sets (in mathematics), an empty array should yield nothing. Alternatively, we could consider supporting values like:
- `[null]`
- `[null, value-1]`

By doing so, we can treat null on par with other values and seamlessly combine it with them. The downside is that it assumes `"null"` is not among of the possible options, but perhaps you can come up with a workaround for that :)

Reply:
I get what you're saying, but it aligns with what we've agreed on in the FE/BE spec for this, so I'm tempted to stick with it. It's true that it doesn't respect a pure mathematical set interpretation, but it does make somewhat sense to filter on things that have no value with an empty array.

So going to stick with it as well, but thanks for challenging :)

Reply:
Yes, that's a good point, and might be something we will need sooner rather than later, actually.

I'll remember this when we get to it.